gateway/src/inference/providers/openai.rs

```diff
@@ -382,23 +398,68 @@ pub(super) fn prepare_openai_tools<'a>(
     }
 }
 
-fn tensorzero_to_openai_system_message(system: Option<&str>) -> Option<OpenAIRequestMessage<'_>> {
-    system.map(|instructions| {
-        OpenAIRequestMessage::System(OpenAISystemRequestMessage {
-            content: instructions,
-        })
-    })
+/// This function is complicated only by the fact that OpenAI and Azure require
+/// different instructions depending on the json mode and the content of the messages.
+///
+/// If ModelInferenceRequestJsonMode::On and the system message or instructions does not contain "JSON"
+/// the request will return an error.
+/// So, we need to format the instructions to include "Respond using JSON." if it doesn't already.
+fn tensorzero_to_openai_system_message<'a>(
+    system: Option<&'a str>,
+    json_mode: &ModelInferenceRequestJsonMode,
+    messages: &[OpenAIRequestMessage<'a>],
+) -> Option<OpenAIRequestMessage<'a>> {
+    match system {
+        Some(instructions) => {
+            match json_mode {
+                ModelInferenceRequestJsonMode::On => {
+                    if messages.iter().any(|msg| {
+                        msg.content()
+                            .map(|c| c.to_lowercase().contains("json"))
+                            .unwrap_or(false)
+                    }) || instructions.to_lowercase().contains("json")
+                    {
+                        OpenAIRequestMessage::System(OpenAISystemRequestMessage {
+                            content: Cow::Borrowed(instructions),
+                        })
+                    } else {
+                        let formatted_instructions =
+                            format!("{} Respond using JSON.", instructions);
```

Comment:
I think we should do `Respond using JSON.\n\n{instructions}`, in case there's something weird about the instructions.

Also you mix "system" and "instructions" but we don't use "instructions" elsewhere? `Some(system)` would be clearer.